Create new module from overleaf/filestore persistors (#1)
overleaf/object-persistor: README.md

# @overleaf/object-persistor

Stores arbitrary objects in multiple backends, with support for falling back to a secondary backend if the object can't be found in the primary.

> Contains a workaround within the GCS backend to allowlifecycle rules to keep objects for a set period of time from deletion, which can't currently be accomplished with GCS's own rules. (See configuration-specific notes later)

Reviewer: ```suggestion
Contains a workaround within the GCS backend to allow lifecycle rules to keep objects for a set period of time from deletion, which can't currently be accomplished with GCS's own rules. (See configuration-specific notes later)
```

## Backends available

- S3
- GCS
- Filesystem (FS)

## Getting started

```JavaScript
const ObjectPersistor = require('object-persistor')

const config = {
  // see 'Configuration' section below
}
const OP = ObjectPersistor(config)
// to use the promise-based API specifically
const persistor = OP.promises
```

## Persistor API

> The object returned by `ObjectPersistor()` has a callback API and a promise-based API (under `.promises`). This documentation covers the promises API. For the callback API, the standard convention of a final parameter with either an error as the first parameter and return value as the second applies.

Reviewer: I think it's worth exposing only the promises API and leave the promise/callback dance to whoever will consume that module. It would make it easier to figure out when we can drop the callback API.

```JavaScript
const ObjectPersistor = require('object-persistor')
const { Errors } = ObjectPersistor
```

### Methods

#### sendStream

```JavaScript
async function sendStream(bucketName, key, readStream, sourceMd5 = null)
```

Uploads a stream to the backend.

- `bucketName`: The name of the bucket to upload to
- `key`: The key for the uploaded object
- `readStream`: The data stream to upload
- `sourceMd5` (optional): The md5 hash of the source data, if known. The uploaded data will be compared against this and the operation will fail if it does not match. If omitted, the md5 is calculated as the data is uploaded instead, and verified against the backend.

##### Notes

When using a secondary persistor, this method uploads only to the primary.

If an object already exists at the specified key, it will be overwritten.

#### getObjectStream

```JavaScript
async function getObjectStream(bucketName, key, opts = {})
```

Retrieves a stream from the backend, for reading

- `bucketName`: The name of the bucket to download from
- `key`: The key for the object
- `opts` (optional):

>   - `start`, `end`: Downloads a byte range from the object. Specify both `start` and `end`. `end` is inclusive.

Reviewer: I'd make the option names a bit more explicit. For example: `startByte`, `endByte`.

Author: I think it's better to stick with `start` and `end` as this is what they're called by various APIs. For example, in the HTTP `Range` header they are named like this, as are the options passed to `createReadStream` in both `fs` and the GCS APIs.

##### Returns

> A `stream.ReadStream` to read the data.

Reviewer: ```suggestion
A `stream.Readable` to read the data.
```

#### getObjectSize

```JavaScript
async function getObjectSize(bucketName, key)
```

Returns the size of the stored data

- `bucketName`: The name of the bucket to download from
- `key`: The key for the object

##### Returns

An integer containing the size, in bytes.

##### Notes

When using a secondary persistor this method returns the size from the secondary persistor, if not found on the primary.

#### getObjectMd5Hash

```JavaScript
async function getObjectMd5Hash(bucketName, key)
```

Returns the MD5 hash of the stored data

- `bucketName`: The name of the bucket to download from
- `key`: The key for the object

##### Returns

A string containing the hex representation of the MD5 hash

##### Notes

When using a secondary persistor this method returns the hash from the secondary persistor, if not found on the primary.

#### deleteFile

```JavaScript
async function deleteFile(bucketName, key)
```

Deletes an object

- `bucketName`: The name of the bucket to delete from
- `key`: The key for the object

##### Notes

When using a secondary persistor, this deletes the object from _both_ persistors.

#### deleteDirectory

```JavaScript

> async function deleteDirectory(bucketName, key)

Reviewer: I see that you reverted to the File/Directory nomenclature because you needed to differentiate between methods that operate on a single object and methods that operate on "directories". How about using Object/Prefix instead? (e.g. `deleteObject()` vs `deletePrefix()`) Or use File/Directory for all methods (e.g. `getFile()`, ...)

Reviewer: Disregard this and the next comments I'll mark as resolved :)

```

Uploads a file from the local disk.

- `bucketName`: The name of the bucket to upload to
- `key`: The key for the uploaded object
- `fsPath`: The path on disk to the file for uploading

##### Notes

When using a secondary persistor, this method uploads only to the primary.

If an object already exists at the specified key, it will be overwritten.

This method is designed for applications which may write temporary data out to the disk before uploading.

## Configuration

An object with the relevant configuration should be passed to the main function returned from the module. The object contains both common and backend-specific parameters.

### Common parameters

- `backend` (required): String specifying the primary persistor to use as the storage backend. Must be one of `s3`, `gcs` or `fs`.
- `signedUrlExpiryInMs`: Time before expiry (in milliseconds) of signed URLs

### FS-specific parameters

- `path.uploadFolder` (required): Location for temporary files that are being uploaded

#### Notes

For the `FS` persistor, the `bucketName` should be the full path to the folder on disk where the files are stored.

### S3-specific parameters

> - `s3.key` (required): The AWS access key ID
> - `s3.secret` (required): The AWS secret access key

Reviewer: maybe use the full parameter names from AWS for more legibility:
```suggestion
- `s3.accessKeyId` (required): The AWS access key ID
- `s3.secretAccessKey` (required): The AWS secret access key
```

- `s3.partSize`: The part size for S3 uploads. Defaults to 100 megabytes.
- `s3.endpoint`: For testing - overrides the S3 endpoint to use a different service (e.g. a fake S3 server)
- `s3.pathStyle`: For testing - use old path-style URLs, for services that do not support subdomain-based access

- `s3BucketCreds`: A JSON-encoded string specifying different S3 credentials for accessing different buckets, in the following format. These credentials override the default ones configured in the main `s3` settings:

```json
{
  "bucketName": {

>     "auth_key": "your aws access key ID",
>     "auth_secret": "your aws secret access key"

Reviewer: same here, for consistency
```suggestion
    accessKeyId: "your aws access key ID",
    secretAccessKey: "your aws secret access key"
```

### GCS-specific parameters

GCS authentication is configured automatically via the local service account, or the `GOOGLE_APPLICATION_CREDENTIALS` environment variable.

- `gcs.unlockBeforeDelete`: unlock an event-based hold before deleting. default false (see notes)
- `gcs.deletedBucketSuffix`: if present, copy the object to a bucket with this suffix before deletion (see notes)
- `gcs.deleteConcurrency`: when recursively deleting a directory, the maximum number of delete requests that will be used at once (default 50)
- `gcs.endpoint.apiEndpoint`: For testing - specify a different GCS endpoint to use
- `gcs.endpoint.apiScheme`: For testing - specify a scheme to use for the GCS endpoint (`http` or `https`)
- `gcs.endpoint.projectId`: For testing - the GCS project ID to supply to the overridden backend

#### Notes

In order to support deletion after a period, the GCS persistor allows usage of a two-bucket system. The main bucket contains the live objects, and on delete the objects are first copied to a 'deleted' bucket, and then deleted from the main one. The 'deleted' bucket is then expected to have a lifecycle policy applied to delete objects after a set period.

> In order to facilitate accidental deletion from outside this mechanism, an event-based-hold can be applied by default on the main bucket. This will be unlocked _after_ the object has been copied to the 'deleted' bucket so that the object can then be deleted from the main bucket.

Reviewer: ```suggestion
In order to prevent accidental deletion from outside this mechanism, an event-based-hold can be applied by default on the main bucket. This will be unlocked _after_ the object has been copied to the 'deleted' bucket so that the object can then be deleted from the main bucket.
```
